### ISSUE TITLE
Lage Docker-image med API for ValidateSpellCheckEpub

### ISSUE THREAD

**Issue by josteinaj** [nlbdev/ValidateSpellCheckEpub#1]
Lage et API hvor man kan sende inn en EPUB, og få ut en rapport. Antageligvis bør denne koden være en egen funksjon sånn at man kan kjøre en bok om gangen:

https://github.com/nlbdev/ValidateSpellCheckEpub/blob/8df8838d138d96b40c6214ad3d51b338925eda5a/ValidateSpellCheckEpub/Check.cs#L174-L213

---

Vi hadde et møte hvor vi diskuterte krav (se #2).

## Generelle krav:

- sjekke de samme tingene som den vi har nå
    - sjekker at `lang` (og `xml:lang`?) har riktig språkkode
    - sjekker at strukturen er riktig (hva ligger i dette egentlig?)
    - sjekker om det er brukt noen uvanlige unicode-tegn
    - lager oversikt over hvilke attributter og elementer som er brukt
    - har stavekontroll (både norsk og engelsk?) med en ordliste hvor man kan legge til ord etterhvert
- støtte en zippet EPUB (altså en fil) som input (som i dag)
- støtte en HTML-fil som input
- støtte en unzippet EPUB (altså en mappe) som input
- når en EPUB (zippet eller unzippet) er input:
    - sjekk filen `META-INF/container.xml`, den peker på OPF-fila
    - gå gjennom `<manifest>`-elementet i OPF-fila for å finne alle HTML-filer
- kunne kjøres fra kommandolinja som i dag (se beskrivelse under)
- kunne kjøres som en tjeneste med API i Docker (se beskrivelse under)

## Krav for kommandolinje:

- Bør være en EXE-fil som ikke har tilhørende DLL-filer og ikke trenger noen installasjon (en "portabel EXE-fil")
- Dra-og-slipp en EPUB (zippet eller unzippet) eller en HTML-fil på EXE-fila (er det sånn det virker i dag, med dra-og-slipp?)
- Rapporten kan være en HTML-fil som legges i samme mappe som EXE-fila ligger i, og ha navnet `[input-filnavn] - logisk sjekk.html`

## Krav for API i Docker:

- ordbokfilene lastes ned når tjenesten starter

### Endepunkt: `GET /health`

- returnerer HTTP 503 Service Unavailable hvis ordbok-filene ikke er lastet ned fra Github
- returnerer HTTP 200 hvis ordbok-filene er lastet ned fra Github

### Endepunkt: `POST /check` (se beskrivelse under)

- hvis ordbøkene er mer enn en time gamle: last ned en ny versjon fra Github

#### Query-parametere:

- `path` (optional): absolutt sti til EPUB (unzippet eller zippet) eller HTML-fil

#### Body (optional):

- en EPUB (kun zippet, ikke unzippet) eller en HTML-fil

#### Response:

- Hvis hverken `path` eller body er definert: HTTP 400 Bad Request
- Hvis både `path` eller body er definert: HTTP 400 Bad Request
- Hvis `path` er definert men filen eller mappen ikke finnes: HTTP 404 Bad Request
- Ellers, HTTP 200 OK med:
    - rapporten som HTML i body
    - Content-Type: text/html i header

**Comment by josteinaj** [comment]
Bruke zippet EPUB som input. Nordisk EPUB, ikke NLBPUB.

**Comment by josteinaj** [comment]
1. [lage en liten spesifikasjon først, for å definere hva dette verktøyet skal gjøre helt konkret](https://github.com/nlbdev/ValidateSpellCheckEpub/issues/2)
2. implementere på nytt i .NET 7
3. støtte både API i Docker, og CLI i Windows
4. POST for å sende inn HTML-fil for bruk fra produksjonssystemet
5. POST for å sende inn EPUB-fil (kanskje ikke nødvendig)
6. CLI for å sende inn EPUB-fil for å kjøre i Windows på produsent-PCer

**Comment by josteinaj** [comment]
@usama49 jeg har lagt inn krav på toppen av denne saken. Ser det ok ut?

CC @eliseaas

**Comment by eliseaas** [comment]
Ser bra ut. Bare et par kommentarer.

> sjekker at strukturen er riktig (hva ligger i dette egentlig?)

TOC-struktur, nivå h1-h3

> har stavekontroll (både norsk og engelsk?) med en ordliste hvor man kan legge til ord etterhvert

Både norsk og engelsk støttet i dag. Norsk språkkode som støttes i dag er "no", kan det legges inn støtte for bokmål- og nynorskspesifikke språkkoder i tillegg til "no"? 

> Dra-og-slipp en EPUB (zippet eller unzippet) eller en HTML-fil på EXE-fila (er det sånn det virker i dag, med dra-og-slipp?)

Ikke helt sånn det fungerer i dag. Vi legger epub-filene i en mappe og kjører deretter EXE-fila

**Comment by josteinaj** [comment]
> TOC-struktur, nivå h1-h3

Dette tror jeg sjekkes av nordisk validering allerede? Eller er det noe den ikke fanger opp?

> kan det legges inn støtte for bokmål- og nynorskspesifikke språkkoder i tillegg til "no"?

Det bør vi kunne få til. I første omgang samme ordliste for bokmål og nynorsk, men at språkkodene for bokmål og nynorsk håndteres på samme måte som "no".

> Ikke helt sånn det fungerer i dag. Vi legger epub-filene i en mappe og kjører deretter EXE-fila

Ok, det er for så vidt like greit :+1:.

**Comment by eliseaas** [comment]
> > TOC-struktur, nivå h1-h3
> 
> Dette tror jeg sjekkes av nordisk validering allerede? Eller er det noe den ikke fanger opp?

Ja, den fanger vel opp det samme, men at logisk sjekk viser oss strukturen i samme oversikt som alt annet gjør det lettere å fange opp feil.
